text/unification.md

```diff
@@ -0,0 +1,268 @@
+
+Pattern unification has various refinements (see e.g. Libal and Miller 2016) such as:
+- expecting distinct variables only among the variables that actually occur in the term to match
+- identifying arguments that are irrefutable patterns over variables at their leaves (e.g. `(x,(y,z))` with those variables (i.e. a form of curryfication) (see PR [#14798](https://github.com/coq/coq/pull/14798))
+
+## Heuristics
+
+Facing the non-determinism between several possible projections and a possible imitation, heuristics can be applied. E.g. to infer the return clause of a `match`, we can give priority to the less dependent solution (see function `Evarconv.choose_less_dependent_instance` for equations of the form `Γ ⊢ ?e@{subst} =?= x` with `x` occurring several times in `subst`).
+
+In addition to the heuristics arbitrating between different projections or between projection and imitation, the following heuristics are used to help to instantiate existential variables about which no particular equations otherwise apply.
+- Canonical structures to solve problems of the form `projᵢ t =?= u` by virtual reduction to `projᵢ t =?= projᵢ v`, that is, concretely, to `t =?= v`, where `v` is a canonical equipment of all components of a tuple from only knowing the value of its projection `u` (see functions `Evarconv.check_conv_record` and `Evarconv.check_record`).
+- Type class inference to solve existential variables of some type by proof search (function `Typeclasses.resolve_typeclasses`).
+- First-order unification: e.g. to resolve `?P ?x =?= Q t` by assuming that the intent is to set `?P := Q` and `?x := t` (see `Evarconv.first_order_unification` and subfunction `first_order` of `Evarconv.evar_conv_x`). 
+
+  It is actually not fully clear to me what are all the ins and outs of first-order unification.
+  - In situations like `Q ?x =?= Q t` with `?x` erasable when reducing `Q ?x`, first-order unification gives a hint on how to infer `?x` and this seems to be a useful heuristics.
+  - In situations like `?P t =?= Q t`, it tells not to care about possible instances of `t` in `Q` and this is more disputable (e.g., it is a priori unexpected that `?P 0 =?= (0=0)` is solved into `?P := fun x => 0=x` rather than into `?P := fun x => x=x`).
+
+- Second-order unification heuristics: e.g. to resolve `?P t =?= u` by selecting a number of occurrences of subterms unifiable to `t` in `u` (see `Evarconv.second_order_matching`). Such a heuristic is central to any form of reasoning by induction. As far as I can recap, it is currently used in Coq only for tactics.
+  - One issue is to recognize subterms: we can assume `t` free of existential variables and reason modulo convertibility, or we can reason using syntactic equality, or syntactic equality discarding non-essential, i.e. reinferable arguments (i.e., morally the implicit arguments).
```

Comment:
I think this should be subsumed by introducing constructors of types with eta.  That is, if we have `?P t =?= u`, we can always instantiate `?P` with `fun x => ?Q`, turning the application into a substitution.  And if we have `fst ?x =?= u` (or `match ?x with pair x y => Q end =?= u`) and primitive projections / eta are turned on, it should always be safe to instantiate `?x` with `(?a, ?b)` (though we probably want to run canonical structures first, and we maybe want to make sure that `u` doesn't reduce to something that structurally matches such a projection/match, because otherwise we may forget information that we want to remember for first-order heuristics)

Reply:
Oh, I see this is mentioned below